esmvalcore/_recipe.py

```diff
@@ -449,41 +453,56 @@ def _update_weighting_settings(settings, variable):
     _exclude_dataset(settings, variable, 'weighting_landsea_fraction')
 
 
+def _update_fx_files(step_name, settings, variable, config_user, fx_vars):
+    """Update settings with mask fx file list or dict."""
+    if not fx_vars:
+        return
+
+    fx_vars = [
+        _get_correct_fx_file(variable, fxvar, config_user)
+        for fxvar in fx_vars
+    ]
+
+    fx_dict = {fx_var[1]['short_name']: fx_var[0] for fx_var in fx_vars}
+    settings['fx_files'] = fx_dict
+    logger.info('Using fx_files: %s for variable %s during step %s',
+                pformat(settings['fx_files']),
+                variable['short_name'],
+                step_name)
+
+
 def _update_fx_settings(settings, variable, config_user):
-    """Find and set the FX mask settings."""
-    msg = f"Using fx files for %s of dataset {variable['dataset']}:\n%s"
-    if 'mask_landsea' in settings:
-        logger.debug('Getting fx mask settings now...')
-        fx_dict = _get_landsea_fraction_fx_dict(variable, config_user)
-        fx_list = [fx_file for fx_file in fx_dict.values() if fx_file]
-        settings['mask_landsea']['fx_files'] = fx_list
-        logger.info(msg, 'land/sea masking', pformat(fx_dict))
-
-    if 'mask_landseaice' in settings:
-        logger.debug('Getting fx mask settings now...')
-        settings['mask_landseaice']['fx_files'] = []
-        fx_files_dict = {
-            'sftgif': _get_correct_fx_file(variable, 'sftgif', config_user)}
-        if fx_files_dict['sftgif']:
-            settings['mask_landseaice']['fx_files'].append(
-                fx_files_dict['sftgif'])
-        logger.info(msg, 'land/sea ice masking', pformat(fx_files_dict))
-
-    if 'weighting_landsea_fraction' in settings:
-        logger.debug("Getting fx files for landsea fraction weighting now...")
-        fx_dict = _get_landsea_fraction_fx_dict(variable, config_user)
-        settings['weighting_landsea_fraction']['fx_files'] = fx_dict
-        logger.info(msg, 'land/sea fraction weighting', pformat(fx_dict))
-
-    for step in ('area_statistics', 'volume_statistics'):
-        if settings.get(step, {}).get('fx_files'):
-            var = dict(variable)
-            var['fx_files'] = settings.get(step, {}).get('fx_files')
-            fx_files_dict = {
-                fxvar: _get_correct_fx_file(variable, fxvar, config_user)
-                for fxvar in var['fx_files']}
-            settings[step]['fx_files'] = fx_files_dict
-            logger.info(msg, step, pformat(fx_files_dict))
+    """Update fx settings depending on the needed method."""
```

Comment:
This function is needlessly complicated

Reply:
hang on, looking at it in a jiffy

Reply:
ok simplified now courtesy to @zklaus last suggestions, happy Bouwe or unhappy Bouwe? :beer:

Reply:
Thanks, much better! I still think the function inside a function `_get_fx_vars_from_attribute` is not really needed, or do you have a particular reason for this?

Reply:
deal with it :laughing: